obspy/core/event/resourceid.py

```diff
@@ -0,0 +1,653 @@
+"""
+obspy.core.event.resourceid - ResourceIdentifier
+================================================
+This module defines the ResourceIdentifier class and associated code.
+"""
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+from future.builtins import *  # NOQA
+from future.utils import native_str
+
+import re
+import warnings
+from contextlib import contextmanager
+from copy import deepcopy
+from uuid import uuid4
+from weakref import WeakKeyDictionary, WeakValueDictionary
+
+
+class _ResourceKey(object):
+    """
+    A private semi-singleton class used to refer id strings to objects.
+
+    Only one instance should be created for each unique string/int, unless
+    all reference are garbage collect, in which case a new instance will be
+    created. Constructing instances through the get_resource_key class method
+    is required to ensure this behavior.
+
+    This class allows the python gc to handle cleanup of the
+    ResourceIdentifier stored class state rather than manual reference
+    counting as was implemented before.
+    """
+    # define a mapping from a resource string to a singleton instance
+    _singleton_cache = WeakValueDictionary()
+
+    def __deepcopy__(self, memodict={}):
+        memodict[id(self)] = self
+        return self
+
+    def __copy__(self):
+        return self
+
+    @classmethod
+    def get_resource_key(cls, unique_id):
+        if unique_id not in _ResourceKey._singleton_cache:
+            single = _ResourceKey()
+            _ResourceKey._singleton_cache[unique_id] = single
+        return _ResourceKey._singleton_cache[unique_id]
+
+
+class _ResourceKeyDescriptor(object):
+    """
+    A private descriptor for initializing _Resource_Key instances.
+    """
+
+    def __init__(self, name, default=None):
+        self.name = name + '__'
+        self.default = default
+
+    def __get__(self, instance, owner):
+        return getattr(instance, self.name, self.default)
+
+    def __set__(self, instance, value):
+        # if an object was passed, use the id of the object for hash
+        if value is not None:
+            if not isinstance(value, (int, str, native_str)):
```

Comment:
Yes, some of the objects used as parents are (rightfully) not hashable. For example:

```python
import obspy.core.event as ev
hash(ev.Event())  # raises TypeError
```
 This little hack gets around it. In this case, we are actually just interested in ensuring the key objects are unique (in memory) so the mutability considerations that normally should guide if an object is hashable are not really applicable.

You bring up a good point about possible collisions, although I think it would have to be deliberate or just a very very bad day. I guess if we are concerned about it I could also check if `value` is an int and cast it to a str or float so that the only way an int gets into the hash table is through this if clause? It may incur a slight hit to performance though.

Reply:
Also, looking at the flow of the program, I would be surprised if an int ever makes it here, but it wouldn't be impossible.

Reply:
> This little hack gets around it. In this case, we are actually just interested in ensuring the key objects are unique (in memory) so the mutability considerations that normally should guide if an object is hashable are not really applicable.

I was more thinking to just always hash `id(value)`. I agree that we don't really have to worry about this minuscule chance and it could still happen with any hash.